src/Engine/Driver.java

```diff
@@ -111,29 +106,33 @@ public static void main(String args[])
   @Override // From KeyListener
   public void keyPressed(KeyEvent arg0)
   {
+    // Store off the old controller/view so we can return to them.
+    IView oldView = gameView.view;
+    
     InputHandler.InputAction action = InputHandler.pressKey(arg0);
 
     if( action != InputHandler.InputAction.NO_ACTION )
     {
       // Pass the action on to the main game controller; if it says it's done, switch to previous state.
-      if(gameController.handleInput(action))
+      if(gameStatus.peek().controller.handleInput(action))
       {
+        gameStatus.pop(); // discard our current controller
+        
+        if( gameStatus.isEmpty() )
+        {
+          // We are done here. I would say clean up data or whatever, but this is Java.
+          System.exit(0);
+        }
+
+        ControlState destination = gameStatus.pop(); // grab the old one
+        
         // Reinstate the previous controller/view.
-        // If the top-level controller returns done, it's time to exit.
-        changeGameState( oldController, oldView );
+        changeGameState( destination.controller, destination.view );
 
-        // We've switched back to these, now make sure we don't do it again (forever).
+        // Clean up our old view
         oldView.cleanup();
```

Comment:
Perhaps do cleanup before calling `changeGameState` (and maybe before `System.exit(0)` just for good practice)?